client/services/ReinforcementGeneratorService.js

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for generating positive reinforcement when a user
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  'ReinforcementObjectFactory', 'TranscriptService',
+  function(ReinforcementObjectFactory, TranscriptService) {
+
+    var getTestsKeyedByTagAndIndexes = function(tests) {
+      var splitTests = {};
+      var inputToOriginalIndex = {};
```

Comment:
I think it's worth adding comments to explain what the keys/values of these two objects are, and why they're needed. The names are reasonable but a bit hard to parse for someone who doesn't have context.